### PR TITLE
Some new configuration options and other small improvements

### PR DESCRIPTION
**Pull request by selfdocumentingcode**
Hello! 

I added a couple of new features that I found useful for my use case and also made some refactoring along the way. 
The PR is rather messy and I haven't cleaned up yet since I wasn't sure if you will consider merging this or not. 
If that turns out to be the case, I can fix it up a bit as well as update the documentation. 

I managed to keep backwards compatibility (I think) and keep the tests running for the most part. I didn't manage to run the Gitea tests, even on the unchanged develop branch. I got a token and everything, but I don't know. Would appreciate some help here if this PR were to be merged.

Anyway, the changes I made are as follows:
Features: 
- Add `commit_template` property to use for "fake" PRs in HYBRID and COMMIT modes. 
Did my best to keep it backwards compatible, but there might be some issues with cached configurations. In the local tests, the way the configuration is loaded and prioritized works a bit differently than it does in "prod", so it's hard to say if it's 100% backwards compatible
- Add optional `categorized_include_empty_content` boolean property to make `categorized_output` category entries include the `empty_content` string as an array item when the array would be empty i.e. no PRs.
- Add optional `mode` property for `category` which makes it possible to define a category that only applies to either real PRs or fake PRs. 
By default it's in `HYBRID` mode which makes it apply to both types of PRs just like before. 

Refactoring:
- Extracted more functions out of the large `buildChangelog` function as well as making the new and existing ones either deal with building the placeholder maps (which I renamed to "template contexts" [inspired by Scriban/Liquid] since I thought too many things were called placeholder something) or deal with "rendering" i.e. substituting tokens in the template strings.

Misc. changes:
- I had to add `plugin:@typescript-eslint/recommended` in order to fix an issue where the built-in `Record<T,U>` type was not being recognized. I have no idea how adding that plugin fixed that, but it did. I can, of course, revert it in this PR if it's unwanted.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -25,25 +25,25 @@ jobs:
...
+
   test:
     if: github.event_name  == 'pull_request'
     runs-on: ubuntu-latest
     steps:
       # the below actions use the local state of the action. please replace `./` with `mikepenz/release-changelog-builder-action@{latest-release}`
       # Showcases how to use the action without a prior checkout
       # Since 3.2.0 the configuration can be provided within the `yml` file
-      - name: "Configuration without Checkout"
+      - name: 'Configuration without Checkout'
```
Comment (mikepenz) [inline .github/workflows/ci.yml:42]: Can you please revert the changes made here?

File: .eslintignore [hunk 2]
```diff
@@ -1,3 +1,4 @@
 dist/
 lib/
-node_modules/
\ No newline at end of file
+node_modules/
+__tests__/
```
Comment (mikepenz) [inline .eslintignore:4]: please revert

File: src/configuration.ts [hunk 3]
```diff
@@ -1,30 +1,26 @@
-import {Extractor, PullConfiguration, Regex, Rule, Sort} from './pr-collector/types'
+import {Extractor, PullConfiguration, Regex, Rule} from './pr-collector/types'
 
 export interface Configuration extends PullConfiguration {
-  max_tags_to_fetch: number
-  max_pull_requests: number
-  max_back_track_time_days: number
-  exclude_merge_branches: string[]
```
Comment (mikepenz) [inline src/configuration.ts:7]: why the removal of these?

**Comment by mikepenz** [comment]
Thank you very much @selfdocumentingcode for the proposal and for the PR. 

I scanned through the changes on a high level, to get a better understanding of the changes. 
And I really like the proposed enhancements. 

That said, would it please be possible to break the PR into smaller PRs please. E.g. have the refactoring in its own PR. 
And then also the individual features in their own PR. 

Additionally, I'd appreciate if you could please revert the unrelated changes, like the updates of the dependencies 
Equally, I do see more changes which should please be reverted:
Screenshot 2024-09-15 at 11 12 38

File: .github/workflows/ci.yml (hunk 1 above)
Comment (selfdocumentingcode) [reply .github/workflows/ci.yml]: Of course. That's how prettier thought to format the file. 
I added an override in for prettier to use double quotes in .yml and re-formatted it. 
It's not exactly reverted, since some quotes were single quotes from before, but it should be more consistent now; I hope you agree with the change.

File: src/configuration.ts (hunk 3 above)
Comment (selfdocumentingcode) [reply src/configuration.ts:7]: They are already inherited from `PullConfiguration`.

File: .eslintignore (hunk 2 above)
Comment (selfdocumentingcode) [reply .eslintignore:4]: I agree and I will revert it, but I am experience issues when eslint is active for the files in `__test__` directory. 
For example:
image

I am using VSCode with the latest version of ESLint. 
Do you happen the know what the issue is?

**Comment by selfdocumentingcode** [comment]
@mikepenz Sounds good! I will start setting up the refactoring PR.

**Comment by selfdocumentingcode** [comment]
@mikepenz I created #1373 with the refactoring changes.

**Comment by mikepenz** [comment]
The split of PR was merged. I believe this one can be closed as a result.
